xmtp/src/client.rs

```diff
@@ -112,8 +112,19 @@ where
         Ok(contacts)
     }
 
-    pub fn create_outbound_session(&self, contact: Contact) -> Result<SessionManager, ClientError> {
-        let olm_session = self.account.create_outbound_session(contact.clone());
+    pub fn get_session(&self, contact: &Contact) -> Result<SessionManager, ClientError> {
```

Comment:
Whats the model here? I imagined that sessions where symmetric, in that the origin does not matter when fetching. 

Logic looks like:
- Is there a Session (of any type between ThisInstallation and InstallationA? return
  - if not create an outgoing session
  
 Is there a case where the client handles inbound and outbound sessions differently?

Reply:
Any time you receive a PreKey message you need to handle two cases:

1. This is the first, and we need to create a new session in the DB
2. This is the Nth PreKey message, we should re-use the existing session

Reply:
Appreciate the clarification. I think I had some incorrect assumptions that new sessions were only ever created/initialized when an inbound invite was received. ie. there was a tight coupling between invites and session re-initialization after a decryption state corruption. 

What I am hearing is that 
1) client ought to expose a 'get_existing_session' method rather than 'get_or_create_session'
2) The database will only contain a single session for an installation_id, old sessions can be discarded immediately 

Did I get that right?

Reply:
> new inbound sessions were only ever created/initialized when an inbound invite was received

I think that is correct. But if my Installation A starts two conversations with two other parties, we want both invites to be relayed to my Installation B on the same session.

If Installation B hasn't sent any messages in the other direction yet, both of them will be PreKey messages.

> The database will only contain a single session for an installation_id, old sessions can be discarded immediately
That _feels_ right. We may have to revisit this when we get into the weeds of decryption state corruption. Maybe there are some cases where we want to keep the old session around purely for handling messages that were delivered out of order or something.

Reply:
> we want both invites to be relayed to my Installation B on the same session.

I think this is where I don't follow. If we assume that sessions are channels between installations and conversations are channels between accounts: Does the session information need to be replicated to InstallB? The prekey of one pairwise session would be useless to an other install which would need to establish its own session using unique prekeys. 

I had assumed that the only packet that needed to be relayed is the Conversation::message -- which is the exact message that was sent to the 2 foreign installations. 

For clarity this is my mental model, @neekolas anything standout as wrong/problematic?

At the conversation level the sequence diagram looks like
```mermaid
   sequenceDiagram
     participant A1 as Amal
    participant B1 as Bola
    Note over A1: Amal sends invite
    A1->>B1: Invite(Amal, Bola)
  

```

At the Installation/sesion level
```mermaid
sequenceDiagram
    participant A2 as Amal2
     participant A1 as Amal1
    participant B1 as Bola1
    participant B2 as Bola2
    Note over A1: Amal1 sends invite
    par
       A1->>B1: Invite(Amal, Bola)
    and    
       A1->>B2: Invite(Amal, Bola)
and  
    A1 ->> A2: Gossip(Invite(Amal,Bola))
    end
   Note over A2: **Amal2 initiates sessions if needed
```

- High level messages addressed between Accounts are fanned out to each installation. e.g ConversationInvite
- Session state does not need to be relayed. Amal2 is oblivious to the sessions of Amal1

** open question of whether Amal2 should pre-emptively initialize sessions or handle it lazily.

Reply:
Everything in that diagram looks correct to me. I'm not suggesting we gossip the sessions between Amal1 and Bola1 or Bola2 or anything like that.

I'm thinking about the much narrower case of subsequent conversations being created by Amal1 (talking to Charlie or Daniel) and then gossipped to Amal2. The Amal1 <> Amal2 session should be the same between all subsequent invites.

> ** open question of whether Amal2 should pre-emptively initialize sessions or handle it lazily.

I don't have strong opinions there. Leaning towards lazily just because it will reduce some unnecessary work if Amal2 never actually needs to publish.